eclair-core/src/main/scala/fr/acinq/eclair/Setup.scala

```diff
@@ -65,12 +66,15 @@ import scala.util.{Failure, Success}
  *
  * Created by PM on 25/01/2016.
  *
- * @param datadir  directory where eclair-core will write/read its data.
- * @param seed_opt optional seed, if set eclair will use it instead of generating one and won't create a seed.dat file.
- * @param db       optional databases to use, if not set eclair will create the necessary databases
+ * @param datadir           directory where eclair-core will write/read its data.
+ * @param pluginParams      parameters for all configured plugins.
+ * @param pluginCommitments plugin-specific data types which contain an implementation of AbstractCommitments trait.
+ * @param seed_opt          optional seed, if set eclair will use it instead of generating one and won't create a seed.dat file.
+ * @param db                optional databases to use, if not set eclair will create the necessary databases
  */
 class Setup(datadir: File,
             pluginParams: Seq[PluginParams],
+            pluginCommitments: Seq[HasAbstractCommitments],
```

Comment:
> I think we shouldn't need to update this file at all.

All requirements seem doable except this one. I think there's no way around doing _something_ in `Setup` when plugin is enriched with new features because `Setup` is the only point through which plugin can influence a system at all (not even a plugin itself but its data since `Plugin` is located in `eclair-node` and not visible, I think intentionally, to `eclair-core` where `Setup` resides).

Reply:
It depends, I think that's something that could maybe be added somehow to `PluginParams`?
This way plugins that want to expose custom payments to the `PostRestart` can return them via this mechanism?
It doesn't touch the `Setup` at all, just requires `PostRestartHtlcCleaner` to fetch more `brokenHtlcs` from plugins (details to bikeshed).
Do you think that could work?

Reply:
I'm thinking about something like this:

- a trait which declares post-restart logic and has default core implementation, an instance of it is to be used in `PostRestartHtlcCleaner`.

- htlc-enabled plugins provide their own instances which build on top of base one, these can even be combined if we have many such plugins (to collect all HTLCs from all plugins before processing).

- these custom implementation instances can be supplied via `PluginParams` but `Setup` still has to send a resulting replacement instance to `PostRestartHtlcCleaner`. That's still a change in `Setup`, but less change than there is currently.

Reply:
I like that direction, but I'd like to make it even more transparent by putting more responsibility and flexibility on the plugins.
What about replacing the current `case class PluginParams` with something like that:

```
/** Custom plugin parameters. */
sealed trait PluginParams {
  /** Plugin's friendly name. */
  def name: String
}

/** Parameters for a plugin that adds support for an experimental or unofficial Bolt9 feature. */
sealed trait CustomFeaturePlugin extends PluginParams {
  /** A set of LightningMessage tags that the plugin wants to listen to. */
  def messageTags: Set[Int]

  /** Feature bit that the plugin wants to advertise through Init message. */
  def feature: Feature
}

/** Parameters for a plugin that defines custom commitment transactions (or non-standard HTLCs). */
sealed trait CustomCommitmentsPlugin extends PluginParams {
  /**
   * If we do nothing after a restart, incoming HTLCs that were committed upstream but not relayed will eventually
   * expire. If your plugin defines non-standard HTLCs, and they need to be automatically failed, they should be
   * returned by this method.
   *
   * Outgoing HTLC sets that are still pending may either succeed or fail: we need to watch them to properly forward the
   * result upstream to preserve channels. If you have non-standard HTLCs that may be in this situation, they should be
   * returned by this method.
   */
  def getBrokenHtlcs: BrokenHtlcs
}
```

Then your plugin defines its params depending on the type of plugin it is, and extends the right traits.
We simply need to update the PostRestartHtlcCleaner to collect all plugins that are `CommitmentsPluginParams` and call their `getBrokenHtlcs` method to add to the list it finds by default. How does that sound? I think this way of defining what type of capabilities plugins can have through traits is nice and flexible.

Reply:
Yes, it sounds good, should be easier to implement what I had in mind. I'll send another PR, then :)